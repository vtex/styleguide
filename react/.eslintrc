{
  "extends": "vtex-react",
  "env": {
    "browser": true,
    "es6": true,
    "jest": true
  }
}
