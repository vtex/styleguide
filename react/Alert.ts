import Alert from './components/Alert/index'

export default Alert
