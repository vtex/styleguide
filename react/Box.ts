import Box from './components/Box/index'

export default Box
