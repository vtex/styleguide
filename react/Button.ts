import Button from './components/Button/index'

export default Button
