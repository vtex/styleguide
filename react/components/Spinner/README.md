#### A Spinner is a way of showing the user something is being loaded, either because of a user action or not. It should be used for short, indeterminate loading times.

### üëç Dos

- Mind the color contrast with the background.
- For very long waiting times (10+ seconds) try giving some feedback about the progress that is being made in the background.
- If you're showing a spinner in a button make sure to make it disabled while loading.
- For very recurrent actions where the probability of success if very high consider using an optimistic approach. E.g.: a message being sent.

### üëé Don'ts

- Only use spinners for loading times of more than 1 second. Less than that it'll cause more harm than good.
- Avoid showing too many spinners at a given time in the same screen, it'll look overwhelming.

### Related components

- If you can assess the progression of the task prefer a <a href="#/Components/Display/Progress">Progress</a>.
- For whole screen loading try using Skeleton Pages (work in progress).

#### Default

```js
<Spinner />
```

#### Custom color and size

```js
<span className="dib c-muted-1">
  <Spinner color="currentColor" size={20} />
</span>
```
