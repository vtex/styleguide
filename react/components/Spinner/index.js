import React from 'react'
import PropTypes from 'prop-types'

import { baseClassname } from '../icon/utils'

const radius = 40
const circ = 2 * radius * Math.PI

class Spinner extends React.Component {
  render() {
    const { color, size, block } = this.props

    return (
      <svg
        className={`${baseClassname('spinner')} ${
          !color ? 'c-action-primary' : ''
        } ${block ? 'db' : ''}`}
        xmlns="http://www.w3.org/2000/svg"
        viewBox="0 0 100 100"
        preserveAspectRatio="xMidYMid"
        height={size}
        width={size}
      >
        <style>
          {`
            @keyframes vtex-spinner-rotate {
              from {
                transform-origin: 50% 50%;
                transform: rotate(0deg);
              }
              to {
                transform-origin: 50% 50%;
                transform: rotate(360deg);
              }
            }

            @keyframes vtex-spinner-fill {
              0% {
                stroke-dasharray: 0 0 2 ${circ};
              }
              50% {
                stroke-dasharray: 0 0 ${circ * 0.75} ${circ};
              }
              100% {
                stroke-dasharray: 0 ${circ - 2} ${circ * 0.75} ${circ};
              }
            }

            .vtex-spinner_circle {
              animation: vtex-spinner-fill 1.25s infinite cubic-bezier(0.455, 0.030, 0.515, 0.955), vtex-spinner-rotate 0.625s infinite linear;
            }
          `}
        </style>

        <circle
          className="vtex-spinner_circle"
          cx="50"
          cy="50"
          fill="none"
          r={radius}
          stroke={color || 'currentColor'}
          strokeWidth="10"
          strokeDasharray={`0 0 ${circ * 0.75} ${circ}`}
          strokeLinecap="round"
          strokeDashoffset="1"
        />
      </svg>
    )
  }
}

Spinner.propTypes = {
  /** Color of the spinner */
  color: PropTypes.string,
  /** Size (diameter) of the spinner */
  size: PropTypes.number,
  /** Sets the display to block */
  block: PropTypes.bool,
}

Spinner.defaultProps = {
  block: false,
  size: 40,
}

export default Spinner
