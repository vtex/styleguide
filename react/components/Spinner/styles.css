@keyframes vtex-spinner-rotate {
  from {
    transform: translate3d(0,0,0) rotate(0deg);
  }
  to {
    transform: translate3d(0,0,0) rotate(360deg);
  }
}

.rotate {
  transform-origin: 50% 50%;
  animation: vtex-spinner-rotate 0.5s infinite linear;
}
