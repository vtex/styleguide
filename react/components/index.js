export { default as ActionMenu } from './ActionMenu'
export { default as Alert } from './Alert'
export { default as AutocompleteInput } from './AutocompleteInput'
export { default as BarChart } from './EXPERIMENTAL_Charts/BarChart'
export { default as Box } from './Box'
export { default as BubbleChart } from './EXPERIMENTAL_Charts/BubbleChart'
export { default as Button } from './Button'
export { default as ButtonGroup } from './ButtonGroup'
export { default as ButtonPlain } from './ButtonPlain'
export { default as ButtonWithIcon } from './ButtonWithIcon'
export { default as Card } from './Card'
export { default as Checkbox } from './Checkbox'
export { default as CheckboxGroup } from './CheckboxGroup'
export { default as Collapsible } from './Collapsible'
export { default as ColorPicker } from './ColorPicker'
export { default as Conditions } from './Conditions'
export { default as DatePicker } from './DatePicker'
export { default as Divider } from './Divider'
export { default as Dropdown } from './Dropdown'
export { default as Dropzone } from './Dropzone'
export { default as EmptyState } from './EmptyState'
export { default as EXPERIMENTAL_Conditions } from './EXPERIMENTAL_Conditions'
export { default as EXPERIMENTAL_Select } from './EXPERIMENTAL_Select'
export { default as EXPERIMENTAL_Table } from './EXPERIMENTAL_Table'
export { default as EXPERIMENTAL_useCheckboxTree } from './EXPERIMENTAL_useCheckboxTree'
export { default as EXPERIMENTAL_useTableMeasures } from './EXPERIMENTAL_Table/hooks/useTableMeasures'
export { default as EXPERIMENTAL_useTableProportion } from './EXPERIMENTAL_Table/hooks/useTableProportion'
export { default as EXPERIMENTAL_useTableSort } from './EXPERIMENTAL_Table/hooks/useTableSort'
export { default as EXPERIMENTAL_useTableVisibility } from './EXPERIMENTAL_Table/hooks/useTableVisibility'
export { default as FilterBar } from './FilterBar'
export { default as FilterOptions } from './FilterOptions'
export { default as FloatingActionBar } from './FloatingActionBar'
export { default as IconArrowBack } from './icon/ArrowBack'
export { default as IconArrowDown } from './icon/ArrowDown'
export { default as IconArrowUp } from './icon/ArrowUp'
export { default as IconBars } from './icon/Bars'
export { default as IconBold } from './icon/Bold'
export { default as IconCalendar } from './icon/Calendar'
export { default as IconCaretDown } from './icon/CaretDown'
export { default as IconCaretLeft } from './icon/CaretLeft'
export { default as IconCaretRight } from './icon/CaretRight'
export { default as IconCaretUp } from './icon/CaretUp'
export { default as IconCheck } from './icon/Check'
export { default as IconCheckPartial } from './icon/CheckPartial' // not a render entrypoint
export { default as IconClear } from './icon/Clear'
export { default as IconClock } from './icon/Clock'
export { default as IconClose } from './icon/Close'
export { default as IconCog } from './icon/Cog'
export { default as IconColumns } from './icon/Columns'
export { default as IconCopy } from './icon/Copy'
export { default as IconDelete } from './icon/Delete'
export { default as IconDensity } from './icon/Density'
export { default as IconDeny } from './icon/Deny'
export { default as IconDownload } from './icon/Download'
export { default as IconEdit } from './icon/Edit'
export { default as IconExternalLink } from './icon/ExternalLink'
export { default as IconExternalLinkMini } from './icon/ExternalLinkMini'
export { default as IconFailure } from './icon/Failure'
export { default as IconFile } from './icon/File'
export { default as IconFilter } from './icon/Filter'
export { default as IconGrid } from './icon/Grid'
export { default as IconHelp } from './icon/Help'
export { default as IconImage } from './icon/Image'
export { default as IconInfo } from './icon/Info'
export { default as IconInlineGrid } from './icon/InlineGrid'
export { default as IconItalic } from './icon/Italic'
export { default as IconLink } from './icon/Link'
export { default as IconLocation } from './icon/Location'
export { default as IconMinus } from './icon/Minus'
export { default as IconOptionsDots } from './icon/OptionsDots'
export { default as IconOrderedList } from './icon/OrderedList'
export { default as IconPlus } from './icon/Plus'
export { default as IconPlusLines } from './icon/PlusLines'
export { default as IconPrinter } from './icon/Printer'
export { default as IconSave } from './icon/Save'
export { default as IconSearch } from './icon/Search'
export { default as IconShoppingCart } from './icon/ShoppingCart'
export { default as IconSuccess } from './icon/Success'
export { default as IconTableFilter } from './icon/TableFilter'
export { default as IconText } from './icon/Text'
export { default as IconUnderline } from './icon/Underline'
export { default as IconUnorderedList } from './icon/UnorderedList'
export { default as IconUpload } from './icon/Upload'
export { default as IconUser } from './icon/User'
export { default as IconVisibilityOff } from './icon/VisibilityOff'
export { default as IconVisibilityOn } from './icon/VisibilityOn'
export { default as IconWarning } from './icon/Warning'
export { default as Input } from './Input'
export { default as InputButton } from './InputButton'
export { default as InputCurrency } from './InputCurrency'
export { default as InputPassword } from './InputPassword'
export { default as InputSearch } from './InputSearch'
export { default as Layout } from './Layout'
export { default as LineChart } from './EXPERIMENTAL_Charts/LineChart'
export { default as Link } from './Link'
export { default as Menu } from './Menu'
export { default as Modal } from './Modal'
export { default as ModalDialog } from './ModalDialog'
export { default as MultiSelect } from './MultiSelect'
export { default as NumericStepper } from './NumericStepper'
export { default as PageBlock } from './PageBlock'
export { default as PageHeader } from './PageHeader'
export { default as Pagination } from './Pagination'
export { default as PasswordInput } from './PasswordInput'
export { default as Progress } from './Progress'
export { default as ProgressBar } from './ProgressBar'
export { default as Radio } from './Radio'
export { default as RadioGroup } from './RadioGroup'
export { default as ScatterChart } from './EXPERIMENTAL_Charts/ScatterChart'
export { default as SelectableCard } from './SelectableCard'
export { default as Slider } from './Slider'
export { default as Spinner } from './Spinner'
export { default as Tab } from './Tab'
export { default as Table } from './Table'
export { default as Tabs } from './Tabs'
export { default as Tag } from './Tag'
export { default as Textarea } from './Textarea'
export { default as TimePicker } from './TimePicker'
export { default as Toggle } from './Toggle'
export { default as Tooltip } from './Tooltip'
export { default as Totalizer } from './Totalizer'
export {
  ToastContext,
  ToastConsumer,
  ToastProvider,
  withToast,
} from './ToastProvider'
