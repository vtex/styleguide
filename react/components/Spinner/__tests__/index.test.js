import React from 'react'
import { render } from '@testing-library/react'

import Spinner from '../index'

describe('Spinner', () => {
  it('matches snapshot', () => {
    const component = render(<Spinner />)
    expect(component.asFragment()).toMatchSnapshot()
  })
  it('renders with the correct color', () => {
    const component = render(<Spinner color="#000" />)
    expect(component.asFragment()).toMatchSnapshot()
    expect(
      component.container
        .querySelector('.vtex-spinner_circle')
        .getAttribute('stroke')
    ).toBe('#000')
  })
  it('renders with the correct size', () => {
    const component = render(<Spinner size={100} />)
    expect(component.asFragment()).toMatchSnapshot()
    expect(
      component.container
        .querySelector('.vtex__icon-spinner')
        .getAttribute('height')
    ).toBe('100')
    expect(
      component.container
        .querySelector('.vtex__icon-spinner')
        .getAttribute('width')
    ).toBe('100')
  })
})
