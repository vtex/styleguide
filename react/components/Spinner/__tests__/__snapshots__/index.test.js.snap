// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Spinner matches snapshot 1`] = `
<DocumentFragment>
  <svg
    class="vtex__icon-spinner  c-action-primary "
    height="40"
    preserveAspectRatio="xMidYMid"
    viewBox="0 0 100 100"
    width="40"
    xmlns="http://www.w3.org/2000/svg"
  >
    <style>
      
            @keyframes vtex-spinner-rotate {
              from {
                transform-origin: 50% 50%;
                transform: rotate(0deg);
              }
              to {
                transform-origin: 50% 50%;
                transform: rotate(360deg);
              }
            }

            @keyframes vtex-spinner-fill {
              0% {
                stroke-dasharray: 0 0 2 251.32741228718345;
              }
              50% {
                stroke-dasharray: 0 0 188.49555921538757 251.32741228718345;
              }
              100% {
                stroke-dasharray: 0 249.32741228718345 188.49555921538757 251.32741228718345;
              }
            }

            .vtex-spinner_circle {
              animation: vtex-spinner-fill 1.25s infinite cubic-bezier(0.455, 0.030, 0.515, 0.955), vtex-spinner-rotate 0.625s infinite linear;
            }
          
    </style>
    <circle
      class="vtex-spinner_circle"
      cx="50"
      cy="50"
      fill="none"
      r="40"
      stroke="currentColor"
      stroke-dasharray="0 0 188.49555921538757 251.32741228718345"
      stroke-dashoffset="1"
      stroke-linecap="round"
      stroke-width="10"
    />
  </svg>
</DocumentFragment>
`;

exports[`Spinner renders with the correct color 1`] = `
<DocumentFragment>
  <svg
    class="vtex__icon-spinner   "
    height="40"
    preserveAspectRatio="xMidYMid"
    viewBox="0 0 100 100"
    width="40"
    xmlns="http://www.w3.org/2000/svg"
  >
    <style>
      
            @keyframes vtex-spinner-rotate {
              from {
                transform-origin: 50% 50%;
                transform: rotate(0deg);
              }
              to {
                transform-origin: 50% 50%;
                transform: rotate(360deg);
              }
            }

            @keyframes vtex-spinner-fill {
              0% {
                stroke-dasharray: 0 0 2 251.32741228718345;
              }
              50% {
                stroke-dasharray: 0 0 188.49555921538757 251.32741228718345;
              }
              100% {
                stroke-dasharray: 0 249.32741228718345 188.49555921538757 251.32741228718345;
              }
            }

            .vtex-spinner_circle {
              animation: vtex-spinner-fill 1.25s infinite cubic-bezier(0.455, 0.030, 0.515, 0.955), vtex-spinner-rotate 0.625s infinite linear;
            }
          
    </style>
    <circle
      class="vtex-spinner_circle"
      cx="50"
      cy="50"
      fill="none"
      r="40"
      stroke="#000"
      stroke-dasharray="0 0 188.49555921538757 251.32741228718345"
      stroke-dashoffset="1"
      stroke-linecap="round"
      stroke-width="10"
    />
  </svg>
</DocumentFragment>
`;

exports[`Spinner renders with the correct size 1`] = `
<DocumentFragment>
  <svg
    class="vtex__icon-spinner  c-action-primary "
    height="100"
    preserveAspectRatio="xMidYMid"
    viewBox="0 0 100 100"
    width="100"
    xmlns="http://www.w3.org/2000/svg"
  >
    <style>
      
            @keyframes vtex-spinner-rotate {
              from {
                transform-origin: 50% 50%;
                transform: rotate(0deg);
              }
              to {
                transform-origin: 50% 50%;
                transform: rotate(360deg);
              }
            }

            @keyframes vtex-spinner-fill {
              0% {
                stroke-dasharray: 0 0 2 251.32741228718345;
              }
              50% {
                stroke-dasharray: 0 0 188.49555921538757 251.32741228718345;
              }
              100% {
                stroke-dasharray: 0 249.32741228718345 188.49555921538757 251.32741228718345;
              }
            }

            .vtex-spinner_circle {
              animation: vtex-spinner-fill 1.25s infinite cubic-bezier(0.455, 0.030, 0.515, 0.955), vtex-spinner-rotate 0.625s infinite linear;
            }
          
    </style>
    <circle
      class="vtex-spinner_circle"
      cx="50"
      cy="50"
      fill="none"
      r="40"
      stroke="currentColor"
      stroke-dasharray="0 0 188.49555921538757 251.32741228718345"
      stroke-dashoffset="1"
      stroke-linecap="round"
      stroke-width="10"
    />
  </svg>
</DocumentFragment>
`;
